### PR TITLE
Make monitorPeerConnection non-blocking

### LINKED ISSUE #4064: Outbound governor does not remove a hot peer when mux errors

### PR DESCRIPTION
**Pull request by coot**
# Description

Fixes #4064

TODO:
* [ ] ~improve network simulation test so it catches the problem~ (it will be done in another PR)
* [x] update changelog

- connection-manager: reworded some comments
- peer-state-actions: make monitorPeerSelection non-blocking
- peer-state-actions: use updateUnlessCold
- peer-state-actions: use epErrorDelay
- exit-policy: derive semigroup via Max
- exit-policy: derive Fractional instance for ReconnectDelay
- peer-selection: trace reconnect delay with 'TraceDemoteAsynchronous'
- peer-selection: refactor monitoring connections
- peer-selection: added localRoots to PeerSelectionCounters

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [x] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs [hunk 1]
```diff
@@ -212,9 +212,9 @@ data PeerSelectionActions peeraddr peerconn m = PeerSelectionActions {
 -- | Callbacks which are performed to change peer state.
 --
 data PeerStateActions peeraddr peerconn m = PeerStateActions {
-    -- | Monitor peer state.
+    -- | Monitor peer state.  Must be non-blocking.
     --
-    monitorPeerConnection    :: peerconn -> STM m (PeerStatus, ReconnectDelay),
+    monitorPeerConnection    :: peerconn -> STM m (PeerStatus, Maybe ReconnectDelay),
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:217]: If it must be non-blocking could we somehow get rid of `STM`?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs [hunk 2]
```diff
@@ -742,27 +742,33 @@ withPeerStateActions PeerStateActionsArguments {
...
         (,) <$> readTVar pchPeerStatus
             <*> (g <$> traverse f pchAppHandles)
       where
         f :: ApplicationHandle muxMode ByteString m a b
-          -> STM m (Map MiniProtocolNum (HasReturned a))
-        f = sequence <=< readTVar . ahMiniProtocolResults
+          -> STM m (Map MiniProtocolNum (Maybe (HasReturned a)))
+             -- do not block when a mini-protocol is still running
+        f =  traverse (((\stm -> (Just <$> stm) `orElse` pure Nothing)))
+         <=< readTVar . ahMiniProtocolResults
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs:761]: I believe this is the culprit? Maybe explain here with more detail why we do not want this to block.

**Review by bolt12 — APPROVED** [review]

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs (hunk 1 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:217]: no because the `Guarded` monoid is executed in `STM` monad.

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:217]: Right, I had edited this comment don't know why it didn't get through. But I think this comment is misleading, we want `monitorPeerConnection` to be non-blocking in a particular function's logic. Saying this at the top level and then have a `STM` type is bit confusing. I'd be more accurate and explain in what situation we do not want to block

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs]: Done

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs [hunk 3]
```diff
@@ -753,6 +760,7 @@ withPeerStateActions PeerStateActionsArguments {
     monitorPeerConnection PeerConnectionHandle { pchPeerStatus, pchAppHandles } =
         (,) <$> readTVar pchPeerStatus
             <*> (g <$> traverse f pchAppHandles)
+        `orElse` throwSTM MonitorPeerConnectionBlocked
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs:763]: Nice!

**Review by bolt12 — APPROVED** [review]

**Comment by coot** [comment]
bors merge

**Pull request merged**